src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -71,51 +71,182 @@ public AssetProvider CreateAssetProvider(PinnedSolutionInfo solutionInfo, Soluti
             Checksum solutionChecksum,
-            bool fromPrimaryBranch,
             int workspaceVersion,
+            bool fromPrimaryBranch,
+            Func<Solution, ValueTask<T>> doWorkAsync,
             CancellationToken cancellationToken)
         {
-            var availableSolution = TryGetAvailableSolution(solutionChecksum);
-            if (availableSolution != null)
-                return availableSolution;
+            // See if anyone else is computing this solution for this checksum.  If so, just piggy-back on that.  No
+            // need for us to force the same computation to happen ourselves.
+            var lazySolution = await GetLazySolutionAndIncrementRefCountAsync().ConfigureAwait(false);
+
+            try
+            {
+                // Actually get the solution, computing it ourselves, or getting the result that another caller was computing.
+                var solution = await lazySolution.GetValueAsync(cancellationToken).ConfigureAwait(false);
+
+                // Store this around so that if another call comes through, they will see the solution we just computed.
+                await SetLastRequestedSolutionAsync(solution).ConfigureAwait(false);
+
+                // Now, pass it to the callback to do the work.  Any other callers into us will be able to benefit from
+                // using this same solution as well
+                var result = await doWorkAsync(solution).ConfigureAwait(false);
+
+                return (solution, result);
 
-            // make sure there is always only one that creates a new solution
-            using (await _availableSolutionsGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+            }
+            finally
             {
-                availableSolution = TryGetAvailableSolution(solutionChecksum);
-                if (availableSolution != null)
-                    return availableSolution;
-
-                var solution = await CreateFullSolution_NoLockAsync(
-                    assetProvider,
-                    solutionChecksum,
-                    fromPrimaryBranch,
-                    workspaceVersion,
-                    CurrentSolution,
-                    cancellationToken).ConfigureAwait(false);
-
-                _lastRequestedSolutionWithChecksum = new(solutionChecksum, solution);
-                return solution;
+                // Now that we're done, update the refcounts for this lazy solution, removing it if the refcount goes back
+                // to zero.
+                await DecrementLazySolutionRefcountAsync().ConfigureAwait(false);
+            }
+
+            async ValueTask<AsyncLazy<Solution>> GetLazySolutionAndIncrementRefCountAsync()
+            {
+                using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+                {
+                    if (_checksumToRefCountAndLazySolution.TryGetValue(solutionChecksum, out var tuple))
```

Comment:
>tuple [](http://example.com/codeflow?start=97&length=5)

`out var (refCount, lazySolution)`

Reply:
not a thing the language supports (yet).